File: CS330/Assignment-5/src/main/java/edu/odu/cs/cs330/items/ItemFactory.java
package edu.odu.cs.cs330.items;

import java.util.Scanner;

/**
 * The Item Creating Wizard
 */
public class ItemFactory {

    /**
     * Name Item Pair 2-tuple(type, model)
     *
     * Simulate a c++ struct by making all attributes public
     */
    private static class ItemPair {
        public String type;  ///< Name of the item to clone
        public Item   model; ///< Model of the item to clone

        /**
         * Default Constructor - Used as sentinel
         */
        public ItemPair()
        {
            type  = null;
            model = null;
        }

        /**
         * Non-Default Constructor
         *
         * @param type the type of anqq item
         * @param item a cloneable item
         */
        public ItemPair(String theType, Item theItem)
        {
            type  = theType;
            model = theItem;
        }
    }

    /**
     * Listing of known items
     */
    private static ItemPair[] knownItems = {
        new ItemPair("Armour", new Armour()),
        new ItemPair("Armor", new Armour()),
        //new ItemPair("Tool", new Tool()),
        new ItemPair("Food", new Consumable()),
        new ItemPair("Potion", new Consumable()),
        new ItemPair("Disposable", new Consumable()),
    };

    /**
     * Create a Item
     *
     * @param type the item to be created
     *
     * @return A item with the specified type
     *     or nullptr if no matching item is found
     */
    public static Item createItem(String type)
    {
        for (ItemPair pair : knownItems) {
            if (pair.type.equals(type)) {
                return pair.model.clone();
            }
        }

        return null;
    }

    /**
     * Determine whether a given item is known
     *
     * @param type the item for which to query
     */
    public static boolean isKnown(String type)
    {
        for (ItemPair pair : knownItems) {
            if (pair.type.equals(type)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Create the appropriate Item class--e.g., Tool, Armour or Consumable.
     *
     * How is **inheritance** used?
     */
    public static Item parseItemLine(Scanner s)
    {
        String keyword = s.next();
        Item item = createItem(keyword);

        // Parse the remainder of the line
        if (item != null) {
            item.read(s);
        }
        else {
            s.nextLine();
        }

        return item;
    }

}

File: CS330/Assignment-4/ItemFactory.cpp
#include <algorithm>

#include "ItemFactory.h"

using std::find_if;

const ItemFactory::ItemPair ItemFactory::_KNOWN_ITEMS[] = {
    {"Armour"    , std::unique_ptr<Item>(new Armour())},
    {"Armor"     , std::unique_ptr<Item>(new Armour())},
    // {"Tool" , std::unique_ptr<Item>(new Tool())}, // TBA in future assignment
    {"Food"      , std::unique_ptr<Item>(new Consumable())},
    {"Potion"    , std::unique_ptr<Item>(new Consumable())},
    {"Disposable", std::unique_ptr<Item>(new Consumable())}
};

//------------------------------------------------------------------------------
Item* ItemFactory::createItem(std::string type)
{
    for (const ItemPair& thePair : _KNOWN_ITEMS) {
        if (thePair.first == type) {
            return (thePair.second)->clone();
        }
    }

    // A item with the given type could not be found
    return nullptr;
}

//------------------------------------------------------------------------------
bool ItemFactory::isKnown(std::string type)
{
    const auto it = find_if(begin(_KNOWN_ITEMS), end(_KNOWN_ITEMS),
                           [type](const ItemPair& thePair) -> bool
                           {
                               return thePair.first == type;
                           });

    return it != end(_KNOWN_ITEMS);
}

//------------------------------------------------------------------------------
/**
 * Read the rest of a line and discard it.
 */
inline
void discardRestOfLine(std::istream& ins)
{
    std::string aether;
    getline(ins, aether);

}

//------------------------------------------------------------------------------
std::istream& operator>>(std::istream& ins, Item*& rd)
{
    std::string type;  // first "word" on line

    // Read the item type and retrieve a template Item from the ItemFactory
    ins >> type;
    rd = ItemFactory::createItem(type);

    if (rd == nullptr) {
        discardRestOfLine(ins);
    }
    else {
        rd->read(ins);
    }

    return ins;
}
